### PR TITLE
Prevent field @timestamp from being removed

### LINKED ISSUE #4370: NoMethodError if field @timestamp is missing
If the field `@timestamp` is missing in an event, the following error occurs:

~~~
NoMethodError: undefined method `to_iso8601' for nil:NilClass
           to_s at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/event.rb:109
         encode at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-codec-line-2.0.2/lib/logstash/codecs/line.rb:53
        receive at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-output-stdout-2.0.3/lib/logstash/outputs/stdout.rb:49
         handle at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/outputs/base.rb:81
    output_func at (eval):47
   outputworker at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/pipeline.rb:276
  start_outputs at /opt/logstash-2.1.0/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/pipeline.rb:193
~~~

**Steps to reproduce**

Start Logstash with the following config:

~~~
input {
  generator {}
}
filter {
  mutate {
    rename => {
      "@timestamp" => "timestamp"
    }
  }
}
output {
  stdout {}
}
~~~

### PR DESCRIPTION
**Pull request by breml**
Fixes #4370

### CODE REVIEW

**Comment by jordansissel** [comment]
@breml overall looks ok to me; I'm not sure if TypeError is the best exception to use for this, though. Thoughts? We can create a new exception type if it makes sense.

**Comment by jordansissel** [comment]
jenkins, test it

**Comment by breml** [comment]
@jordansissel You are right, I just followed the implementation on lIne 127 (https://github.com/u-s-p/logstash/blob/5b24e2f276de755eb1c3b330aa3c50ef848718ce/logstash-core-event/lib/logstash/event.rb#L127), but this is not the same case. I don't know the exception system of ruby well enough to make a sound 	suggestion. I appreciate your help on this.

**Comment by suyograo** [comment]
@breml @jordansissel How about introducing `InvalidOperationException` ? `TypeError` does seem at odds here.

**Comment by breml** [comment]
@suyograo @jordansissel Just updated this PR with InvalidOperationException.

File: logstash-core-event/spec/logstash/event_spec.rb [hunk 1]
```diff
@@ -63,7 +63,7 @@
       it "should raise error when formatting %{+%s} when @timestamp field is missing" do
         str = "hello-%{+%s}"
         subj = subject.clone
-        subj.remove("[@timestamp]")
+        subj.instance_variable_get(:@accessors).del("[@timestamp]")
         expect{ subj.sprintf(str) }.to raise_error(LogStash::Error)
```
Comment (suyograo) [inline logstash-core-event/spec/logstash/event_spec.rb]: Can you check for `InvalidOperationException` here?

**Comment by suyograo** [comment]
@breml can you also add a test where you add a dummy field called `timestamp` not `@timestamp` and assert if you can delete it correctly? Thats a scenrio i've seen in the field, wanted to make sure we protect against that.

**Comment by suyograo** [comment]
@breml ping! I'd love to get this in soon :)

**Comment by breml** [comment]
@suyograo Here you go. (I also have some plugin related issues, where I would love to have some progress soon :)

**Comment by suyograo** [comment]
@breml unfortunately I had to revert this in https://github.com/elastic/logstash/pull/4756 because `rake tests:core` (unit tests) were failing. There are lots of tests checking `@timestamp`. Can you reopen a new PR based on this which has all the tests passing? Thanks much

**Comment by breml** [comment]
@suyograo After several hours of unsuccessful debugging, I think I have found the issue. With https://github.com/elastic/logstash/commit/23dcf467e490e417bf377a0b2a32a73653434879#diff-efbf03f6386a8f9610003e5858706663R15 the change from logstash-core-event to logstash-core-event-java (nicely hidden in a commit about "metrics api") was introduced (intentional?). 

For the first set of errors, the explanation is:
I extended the logstash-event with a new  exception (InvalidOperationException, https://github.com/u-s-p/logstash/commit/c077c9cb66afed26f7855173d24a72abca1d0e0d), this exception is not present in the java event implementation.

For the second set of errors, there seems to be a difference in the behavior between the ruby and the java implementation regarding `subj.instance_variable_get(:@accessors).del("[@timestamp]")`. I assume that `@accessors` is not available in the java implementation.
The reason why I use `instance_variable_get` is, for the respective test case I have to remove the `@timestamp` field, but I can not use the `remove` method, because in the `remove` method it is not allowed to remove the `@timestamp` field.

**I don't think I'm able to resolve this one, as I don't have any knowledge about the new java based event implementation.**

**Comment by purbon** [comment]
@breml turning logstash-core-event-java on in master was done a while ago before we merged the metrics PR, I'm curios why the diff looks like this, there might be something fishy going there for sure.

re: for the new java event, @colinsurprenant or @guyboertje might be able to help here more, any ideas?

**Comment by purbon** [comment]
@breml logstash-core-event-java was enabled back on `Jan 26, 2016`, see https://github.com/elastic/logstash/commit/5a522a00370fb7bd7275d4ae7d7ab996a5fd329e#diff-8b7db4d5cc4b8f6dc8feb7030baa2478, while the commit you point if from `Feb 4`. But I guess the error might be coming out now because in your commit, we actually also updated the Gemfile.lock, that was not updated in the first one. .... :cry: what do you think?

**Comment by breml** [comment]
@purbon At the moment I think, that this is quite frustrating, because the PR was put together in much less time than I needed today to debug this issue. And my PR kind of updates legacy code, which will never be used. This doesn't feel good.

**Comment by purbon** [comment]
@breml I apologies for the inconvenience you're facing, the move to the new java event bring us a ton of new performance, you can see it http://logstash-meter.elastic.co/ for example. This will also enable us to move forward with more persistency and making easier for people to write plugins in any JVM language. Your PR might found a bug in the new java implementation, this is much appreciate it, but I do understand your feelings.

**Comment by breml** [comment]
@purbon Thanks for the link to the metrics. More logstash throughput is always welcome. Hopefully you are right and my PR has found a bug in the new java implementation. In this case it would be a least valuable.

**Comment by suyograo** [comment]
> I don't think I'm able to resolve this one, as I don't have any knowledge about the new java based event implementation.

@breml we can definitely port this to the java implementation. 

> And my PR kind of updates legacy code, which will never be used. This doesn't feel good.
 I wouldn't think about it this way. 2.x release will live for a long time with Ruby implementation of Event, so your change will definitely be used. Also, the same logic will be in Java Event (not able to remove `@timestamp`), so overall your change really benefits us. I'll take another look at the failing tests and Java event related issues.

**Comment by breml** [comment]
@suyograo Thanks for your feedback. Ping me if the Java implementation is ready. I will push the PR again. OK?

**Pull request closed without merging**